pandas/tests/extension/base/reshaping.py

```diff
@@ -173,6 +173,21 @@ def test_merge(self, data, na_value):
                  dtype=data.dtype)})
         self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
 
+    def test_merge_on_extension_array(self, data):
+        # GH 23020
+        df1 = pd.DataFrame({'ext': [1, 2, 3],
+                            'key': data[:3]})
+
+        res = pd.merge(df1, df1, on='key')
+
+        exp = pd.DataFrame(
+            {'key': data[:3],
+             'ext_x': [1, 2, 3],
+             'ext_y': [1, 2, 3]})
+
+        self.assert_frame_equal(res, exp[['ext_x', 'key', 'ext_y']],
+                                check_dtype=True)
+
```

Comment:
It'd be good to have tests that aren't a one-to-one merge. Something like

```python
a, b = data[:2]

key1 = type(data)._from_sequence([a])
key2 = type(data)._from_sequence([a, b])
```


And put those in a DataFrame and do an inner / outer join. LMK if you want help with that.